CRI: add support for dns cluster first policy

feiskyer opened the pull request:
**What this PR does / why we need it**:

PR #29378 introduces ClusterFirstWithHostNet policy but only dockertools was updated to support the feature. 

This PR updates kuberuntime to support it for all runtimes.


**Which issue this PR fixes** 

fixes #43352

**Special notes for your reviewer**:

Candidate for v1.6.

**Release note**:

```release-note
NONE
```

cc @thockin @luxas @vefimova @Random-Liu

feiskyer committed:
CRI: add support for dns cluster first policy

PR #29378 introduces ClusterFirstWithHostNet policy but only dockertools
was updated to support the feature. This PR updates kuberuntime to
support it for all runtimes.

Also fixes #43352.

feiskyer commented:
/assign @thockin @luxas @vefimova @Random-Liu

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/43368)
<!-- Reviewable:end -->

luxas commented:
@feiskyer Thanks!! I'll test and verify in a minute :)

luxas commented:
@feiskyer @thockin This worked for me and solved my issue :+1:, thanks!

I think this is a v1.6.0 candidate then, given the super fast response by @feiskyer 
I said v1.6.1 since I expected it would take some days to get the patch and didn't want to delay the release, but given this is just a merge away, this can be merged now

feiskyer commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

luxas commented:
@k8s-bot cvm gce e2e test this
@k8s-bot gce etcd3 e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot kubemark e2e test this
@k8s-bot non-cri node e2e test this
@k8s-bot node e2e test this

feiskyer committed:
Rewrite resolv.conf for dockershim

PR #29378 introduces ClusterFirstWithHostNet, but docker doesn't support
setting dns options togather with hostnetwork. This commit rewrites
resolv.conf same as dockertools.

feiskyer commented:
@luxas Added a new commit to fix e2e tests. The tests failed because docker v1.12 doesn't support setting dns options togather with host network. But docker v1.13+ does support this.

ncdc commented:
cc @smarterclayton @derekwaynecarr @knobunc @dcbw @danwinship @eparis

luxas commented:
I can confirm that this PR fixes the problem. I hope this will be prioritized and merged into v1.6

thockin commented on pkg/kubelet/dockershim/docker_sandbox.go:
```diff
@@ -92,6 +92,25 @@ func (ds *dockerService) RunPodSandbox(config *runtimeapi.PodSandboxConfig) (str
+		}
+
+		if err := dockertools.RewriteResolvFile(containerInfo.ResolvConfPath, dnsConfig.Servers, dnsConfig.Searches, len(dnsConfig.Options) > 0); err != nil {
+			return createResp.ID, fmt.Errorf("rewrite resolf.conf faield for pod %q: %v", config.Metadata.Name, err)
```
s/resolf/resolv

feiskyer replied:
Ack. Will open another pr to fix this.

thockin commented:
/approve

thockin commented:
/lgtm

thockin commented:
Thanks!

On Mon, Mar 20, 2017 at 10:39 AM, Victoria Efimova <notifications@github.com
> wrote:

> *@vefimova* approved this pull request.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/43368#pullrequestreview-27914124>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFVgVLkVk7U3u4xWbYC2jB79x7vNnoT3ks5rnrnBgaJpZM4MiEf9>
> .
>
